src/Common/src/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs

```diff
@@ -373,37 +478,85 @@ private TValue TryAddOrGetExisting(TValue value, out bool addedValue)
             }
 
             // We've probed to find an empty spot, add to hash
-            if (!TryWriteValueToLocation(value, hashTableLocal, tableIndex))
+            if (!TryWriteSentinelToLocation(hashTableLocal, tableIndex))
             {
+                // After finding the empty spot, it was taken by some other thread
                 Interlocked.Decrement(ref _reserve);
                 return null;
             }
 
             // Now that we've written to the local array, find out if that array has been
             // replaced by expansion. If it has, we need to restart and write to the new array.
-            if (_newHashTable != hashTableLocal)
```

Comment:
I see. The race was here. You could add to old table after resize started and someone could have seen it. If the change was not picked up by resize and another writer, wrote to the new table, you would have effectively two values for the same key.

Reply:
Just to check if I understand the problem/solution correctly - it could be a smaller change to add a `bool published` to every table entry and flip that to true after checking for new table. But that would double the size of the table.

---
In reply to: [356353759](https://github.com/dotnet/corert/pull/7906#discussion_r356353759) [](ancestors = 356353759)